node/pkg/libp2p/libp2p.go

```diff
@@ -239,41 +281,54 @@ func initDHT(ctx context.Context, h host.Host, bootstrap string) *dht.IpfsDHT {
 	}
 	wg.Wait()
 
-	return kademliaDHT
+	return kademliaDHT, nil
+}
+
+func connectToPeers(ctx context.Context, h host.Host, routingDiscovery *drouting.RoutingDiscovery, topicName string) (connected bool, err error) {
+	peerChan, err := routingDiscovery.FindPeers(ctx, topicName)
+	if err != nil {
+		return false, err
+	}
+	var wg sync.WaitGroup
+	for p := range peerChan {
+		if p.ID == h.ID() {
+			continue // No self connection
+		}
+		wg.Add(1)
+		go func(p peer.AddrInfo) {
+			defer wg.Done()
+			err := h.Connect(ctx, p)
+			if err == nil {
+				connected = true
+			}
+		}(p)
+	}
+	wg.Wait()
+	return connected, nil
 }
 
 func DiscoverPeers(ctx context.Context, h host.Host, topicName string, bootstrap string) error {
-	kademliaDHT := initDHT(ctx, h, bootstrap)
+	kademliaDHT, dhtErr := initDHT(ctx, h, bootstrap)
+	if dhtErr != nil {
+		log.Error().Err(dhtErr).Msg("Error initializing DHT")
+		return dhtErr
+	}
 	routingDiscovery := drouting.NewRoutingDiscovery(kademliaDHT)
 	dutil.Advertise(ctx, routingDiscovery, topicName)
 
 	// Look for others who have announced and attempt to connect to them
-	anyConnected := false
-	var wg sync.WaitGroup
-	for !anyConnected {
-		log.Debug().Msg("Searching for peers...")
-		peerChan, err := routingDiscovery.FindPeers(ctx, topicName)
+	var anyConnected bool
+	var err error
+	for i := 0; i < 10 && !anyConnected; i++ {
```

Comment:
Can you please explain the constant 10?

Reply:
it's max retry count, there aren't any reason

Reply:
Aha! This is just suggestion. Since it is just simple retrial, I think there is no need to reinvent a typical for loop:

```go
func DiscoverPeers(ctx context.Context, h host.Host, topicName string, bootstrap string) error {
	defer func() {
		log.Debug().Msg("Peer discovery complete")
	}()

	kademliaDHT, dhtErr := initDHT(ctx, h, bootstrap)
	if dhtErr != nil {
		log.Error().Err(dhtErr).Msg("Error initializing DHT")
		return dhtErr
	}
	routingDiscovery := drouting.NewRoutingDiscovery(kademliaDHT)
	dutil.Advertise(ctx, routingDiscovery, topicName)

	// Look for others who have announced and attempt to connect to them
	const maxConnectionTrials = 10
	for i := 0; i < maxConnectionTrials; i++ {
		log.Debug().Int("connected peers", len(h.Network().Peers())).Msg("Searching for peers...")
		connected, err := connectToPeers(ctx, h, routingDiscovery, topicName)
		if err != nil {
			return err
		}

		if !connected {
			time.Sleep(time.Second * 3) // wait before retrying
			continue
		}

		return nil
	}

	return errors.New("no peers connected")
}
```

Reply:
thank you for the suggestion